File: src/domain/units/conversions.ts
import {
  cubicMetresPerSecond,
  metres,
  type CubicMetresPerSecond,
  type LitresPerSecond,
  type Metres,
  type Millimetres,
} from "./quantities";

/**
 * Hub'Eau renvoie le débit en litres par seconde et la hauteur en millimètres,
 * contrairement à ce que sa documentation laisse croire (C-02, vérifié le
 * 2026-07-30 : `53000.0` valait 53 m³/s).
 *
 * Ce facteur ne doit apparaître NULLE PART ailleurs dans le code. Une double
 * conversion est aussi fausse qu'une conversion oubliée, et bien plus difficile
 * à repérer (BR-002).
 */
const PER_THOUSAND = 1000;

function divide(value: number): number {
  if (!Number.isFinite(value)) {
    throw new RangeError(
      `Conversion impossible : la valeur doit être finie, reçu ${String(value)}.`,
    );
  }
  return value / PER_THOUSAND;
}

/**
 * `null` et `undefined` en entrée donnent `null` en sortie. L'absence de mesure
 * n'est jamais remplacée par zéro : un zéro est un fait mesuré — un assec —,
 * une absence n'en est pas un (BR-007).
 *
 * Une valeur non finie lève plutôt que de rendre `null` : NaN n'est pas « pas
 * de donnée », c'est un bug. Le confondre avec une absence afficherait « la
 * station n'a pas transmis » alors que le code est cassé.
 */
export function toCubicMetresPerSecond(
  value: LitresPerSecond | null | undefined,
): CubicMetresPerSecond | null {
  if (value === null || value === undefined) return null;
  return cubicMetresPerSecond(divide(value));
}

export function toMetres(value: Millimetres | null | undefined): Metres | null {
  if (value === null || value === undefined) return null;
  return metres(divide(value));
}

File: lib/domain/units/conversions.dart
// Conversion l/s → m³/s et mm → m, à un seul endroit (BR-002, C-02).
// Une absence en entree reste une absence en sortie (BR-007) : jamais
// remplacee par zero. Une valeur non finie n'est pas une absence mais un
// defaut, elle leve.

import 'package:martinpecheur/domain/units/quantities.dart';

/// Facteur de conversion entre l/s et m³/s, et entre mm et m. N'apparaît
/// que dans ce fichier (BR-002).
const int _perThousand = 1000;

/// Convertit un débit brut de l'API (l/s) en débit affichable (m³/s).
/// `null` en entrée renvoie `null` (BR-007). Une valeur non finie lève.
CubicMetresPerSecond? toCubicMetresPerSecond(LitresPerSecond? raw) {
  if (raw == null) {
    return null;
  }

  return CubicMetresPerSecond(_divide(raw.value));
}

/// Convertit une hauteur brute de l'API (mm) en hauteur affichable (m).
/// `null` en entrée renvoie `null` (BR-007). Une valeur non finie lève.
Metres? toMetres(Millimetres? raw) {
  if (raw == null) {
    return null;
  }

  return Metres(_divide(raw.value));
}

double _divide(double value) {
  if (!value.isFinite) {
    throw ArgumentError.value(
      value,
      'value',
      'doit être une valeur finie, ni NaN ni infinie',
    );
  }

  return value / _perThousand;
}
